vault/resource_plugin_test.go

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package vault
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+
+	"github.com/hashicorp/terraform-provider-vault/internal/consts"
+	"github.com/hashicorp/terraform-provider-vault/internal/provider"
+	"github.com/hashicorp/terraform-provider-vault/testutil"
+)
+
+const envPluginCommand = "VAULT_PLUGIN_COMMAND"
+
+func TestPlugin(t *testing.T) {
+	const (
+		typ     = "auth"
+		version = "v1.0.0"
+		args    = `["--foo"]`
+		env     = `["FOO=BAR"]`
+
+		argsUpdated = `["--bar"]`
+		envUpdated  = `["FOO=BAZ"]`
+	)
+
+	destName := acctest.RandomWithPrefix("tf/plugin")
+
+	resourceName := "vault_plugin.test"
+	sha256 := strings.Repeat("01234567", 8)
+	sha256Updated := strings.Repeat("12345678", 8)
+	cmd := os.Getenv(envPluginCommand)
+
+	versionsSupported := true
+	if testProvider != nil {
+		m := testProvider.Meta().(*provider.ProviderMeta)
+		versionsSupported = m.GetVaultVersion().GreaterThanOrEqual(provider.VaultVersion112)
+	}
+
+	resource.Test(t, resource.TestCase{
+		ProviderFactories: providerFactories,
+		PreCheck: func() {
+			testutil.TestAccPreCheck(t)
+			testutil.SkipTestEnvUnset(t, envPluginCommand)
+		},
+		Steps: []resource.TestStep{
+			{
+				Config: testPluginConfig(typ, destName, version, sha256, cmd, args, env, versionsSupported),
+				Check: resource.ComposeTestCheckFunc(
+
+					resource.TestCheckResourceAttr(resourceName, consts.FieldType, typ),
+					resource.TestCheckResourceAttr(resourceName, consts.FieldName, destName),
+					testCheckVersionAttr(resourceName, version, versionsSupported),
+					resource.TestCheckResourceAttr(resourceName, fieldSHA256, sha256),
+					resource.TestCheckResourceAttr(resourceName, fieldCommand, cmd),
+					testValidateList(resourceName, fieldArgs, []string{"--foo"}),
+					testValidateList(resourceName, fieldEnv, []string{"FOO=BAR"}),
+				),
+			},
+			{
+				Config: testPluginConfig(typ, destName, version, sha256Updated, cmd, argsUpdated, envUpdated, versionsSupported),
+				Check: resource.ComposeTestCheckFunc(
+
+					resource.TestCheckResourceAttr(resourceName, consts.FieldType, typ),
+					resource.TestCheckResourceAttr(resourceName, consts.FieldName, destName),
+					testCheckVersionAttr(resourceName, version, versionsSupported),
+					resource.TestCheckResourceAttr(resourceName, fieldSHA256, sha256Updated),
+					resource.TestCheckResourceAttr(resourceName, fieldCommand, cmd),
+					testValidateList(resourceName, fieldArgs, []string{"--bar"}),
+					testValidateList(resourceName, fieldEnv, []string{"FOO=BAZ"}),
+				),
+			},
+			testutil.GetImportTestStep(resourceName, false, nil, "env"),
```

Comment:
One thing I don't quite understand, how does this test figure out the correct ID to import with?

Reply:
This particular util sets up a TF Test Step with the `ImportStateVerify` attribute to `true`. When the TF SDK sees a Test Step with this boolean enabled, it imports the TF state and checks against the `resourceName` provided to get the particular resource's ID.

Alternatively, you could also provide a specific `ImportStateID`, but the default behavior is to look for the `resourceName` under the TF state, and get the `id` attribute from it. Hope that makes sense!